ls1intum/Artemis
# `Development`: Remove sub-queries for filtering and ordering posts

### Checklist
#### General


#### Server


### Motivation and Context
While fetching course posts, we previously generated sub-queries to filter and sort posts, in accordance with many variant properties. The use of sub-queries is a bad programming habit, as they drastically reduce performance during database calls.

Also, due to the incompatibility of the DISTINCT and ORDER BY keywords used in Post queries, we previously had to disable server tests related to ordering posts by creation date, vote count, and answer count.

### Description
To remove the sub-queries while preserving the rich filtering and sorting of posts, we introduce three new columns to the post table; resolved, answer_count, and vote_count respectively. By pre-populating these columns for all existing Posts and accounting them with respect to the operations Post go under, we are able to avoid sub-queries and instead utilize the information stored in these columns to realize the filtering and sorting operations of Posts.

To re-enable the previously disabled post-server tests, we remove the DISTINCT keyword and instead use GROUP BY (Post.ID) to achieve the same result.

### Performance Differences
Measurements were performed on a dataset with 237 Posts, 299 Reactions, and 299 AnswerPosts.
| Sorting/Filtering Strategy | Previous Implementation | New Implementation|
|------------|-------:|-----:|
| Unresolved | 774ms | 203ms |
| Sort by Vote Count ASC |  194ms | 168ms |
| Sort by Vote Count DESC | 244ms | 192ms |
| Sort by Answer Count ASC | 174ms | 170ms |
| Sort by Answer Count DESC | 1.07s | 995ms |

[PerformanceMeasurements.zip](https://github.com/ls1intum/Artemis/files/9998019/PerformanceMeasurements.zip)

### Steps for Testing

Prerequisites:
- 1 Tutor
- 1 Course with discussions enabled

1. Log in to Artemis
2. Navigate to Course Communication Page
3. Verify that filters work as before
4. Make sure that posts are ordered correctly with respect to date, votes, and number of answers.
5. Alter votes and number of answers of posts, make sure that they are re-ordered correctly after the changes.
6. Announcement posts are treated as resolved by default. Resolve a non-announcement post and make sure it is not displayed when filtering for unresolved posts.
7. Unresolve the previous post and make sure that it is displayed again while filtering for unresolved posts.
8. Make sure that no Post is displayed more than once. (check Post ID next to Post title)


### Review Progress






#### Code Review
#### Manual Tests


### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| PostService.java | 81% | 97% |
| AnswerPostService.Java | 81% | 79% |
| ReactionService.java | 100% | 100% |
| PostSpecs.java | 87% | 85% |
